new-log-viewer/src/contexts/UrlContextProvider.tsx

```diff
@@ -0,0 +1,254 @@
+        if (null === value) {
+            newHashParams.delete(key);
+        } else {
+            newHashParams.set(key, String(value));
+        }
+    }
+
+    return newHashParams.toString();
+};
+
+/**
+ * Updates search parameters in the current window's URL with the given key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the search parameters. If a value
+ * is `null`, the corresponding kv-pair will be removed from the URL's search parameters.
+ */
+const updateWindowUrlSearchParams = (updates: UrlSearchParamUpdatesType) => {
+    const newUrl = new URL(window.location.href);
+    newUrl.search = getUpdatedSearchParams(updates);
+    window.history.pushState({}, "", newUrl);
+};
+
+/**
+ * Updates hash parameters in the current window's URL with the given key-value pairs.
+ *
+ * @param updates An object containing key-value pairs to update the hash parameters. If a value is
+ * `null`, the corresponding kv-pair will be removed from the URL's hash parameters.
+ */
+const updateWindowUrlHashParams = (updates: UrlHashParamUpdatesType) => {
+    const newHash = getUpdatedHashParams(updates);
+    const currHash = window.location.hash.substring(1);
+    if (newHash === currHash) {
+        return;
+    }
+
+    const newUrl = new URL(window.location.href);
+    newUrl.hash = newHash;
+    window.history.pushState({}, "", newUrl);
+
+    // `history.pushState` doesn't trigger a `hashchange`, so we need to dispatch one manually.
+    window.dispatchEvent(new HashChangeEvent("hashchange"));
+};
+
+/**
+ * Copies the current window's URL to the clipboard. If any `updates` parameters are specified,
+ * the copied URL will include these modifications, but the original window's URL will not be
+ * changed.
+ *
+ * @param searchParamUpdates An object containing key-value pairs to update the search parameters.
+ * If a value is `null`, the corresponding kv-pair will be removed from the URL's search parameters.
+ * @param hashParamsUpdates An object containing key-value pairs to update the hash parameters.
+ * If a value is `null`, the corresponding kv-pair will be removed from the URL's hash parameters.
+ */
+const copyPermalinkToClipboard = (
+    searchParamUpdates: UrlSearchParamUpdatesType,
+    hashParamsUpdates: UrlHashParamUpdatesType,
+) => {
+    const newUrl = new URL(window.location.href);
+    newUrl.search = getUpdatedSearchParams(searchParamUpdates);
+    newUrl.hash = getUpdatedHashParams(hashParamsUpdates);
+    navigator.clipboard.writeText(newUrl.toString())
+        .then(() => {
+            console.log("URL copied to clipboard.");
+        })
+        .catch((error: unknown) => {
+            console.error("Failed to copy URL to clipboard:", error);
+        });
+};
+
+/**
+ * Retrieves all search parameters from the current window's URL.
+ *
+ * @return An object containing the search parameters.
+ */
+const getWindowSearchParams = () => {
```

Comment:
For consistency with the update methods, let's rename these to `getWindowUrlSearchParams` and `getWindowUrlHashParams`.